File: java/Fallback.java
package irresponsible.crouton;

import clojure.lang.IPersistentVector;
import clojure.lang.IPersistentMap;

public class Fallback implements IRoute {
  private final IRoute first;
  private final IRoute second;
  public Fallback(IRoute f, IRoute s) {
    if (null == f || null == s)
      throw new IllegalArgumentException("Both routes provided to Either must not be non-nil");
    first = f;
    second = s;
  }
  public final Object match(IPersistentVector pieces, IPersistentMap places) {
    Object ret = first.match(pieces, places);
    return (ret == null) ? second.match(pieces, places) : ret;
  }
}

File: kotlin/src/Fallback.kt
package irresponsible.crouton

import clojure.lang.IPersistentVector
import clojure.lang.IPersistentMap

class Fallback(private val first: IRoute, private val second: IRoute) : IRoute {
    override fun match(pieces: IPersistentVector, places: IPersistentMap): Any? {
        val ret = first.match(pieces, places)
        return ret ?: second.match(pieces, places)
    }
}
